### PR TITLE
148: ADRs Update

### PR DESCRIPTION
**Pull request by halprin**
# ADRs Update

- An ADR on our AuthN/AuthZ.
- An ADR on how we "wrap" our HAPI FHIR resource classes.

## Issue

#148.

## Checklist

- [x] I have updated the documentation accordingly

### CODE REVIEW

File: adr/012-auth.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+## Context
+
+ReportStream has a form of AuthN/AuthZ that they call "two legged".  It is similar to FHIR's bulk API implementation guide AuthN/AuthZ.  We are inheriting the format.
+
+We chose this because it is exactly what ReportStream does and is accepted practice in the larger FHIR community.  There are a few concepts that ReportStream enforces or uses that we do not because we don't have a requirement for it.  For example, the payload sent to our login endpoint does not require `client_assertion_type`, and we don't send back a `token_expire` since that is encoded into the JWT already.
+
+This form of AuthN/AuthZ goes through multiple steps.
+1. Register the client in TI.  The client sends us a public key and we associate it with that client.  This occurs before any API communication.
+2. When the client wants to send us information through an API call, they will first call our login endpoint with a playload containing their name and a JWT which is signed by their private key.
```
Comment (sfradkin) [inline adr/012-auth.md:21]: `playload` -> `payload`

File: adr/012-auth.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+3. We validate that the JWT is valid given the public key they sent to us earlier.
+4. We generate a short-lived JWT with our own private key and send it back to the client.  This completes the call to the login endpoint.
+5. The client then calls our authenticated endpoints with the JWT we send previously and whatever payload required by the specific endpoint.
```
Comment (sfradkin) [inline adr/012-auth.md:24]: `send` -> `sent`
`payload required` -> `payload is required`

File: adr/012-auth.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+6. We validate that the JWT is valid given our public (or private) key.
+7. With continue with the business of that specific endpoint.
```
Comment (sfradkin) [inline adr/012-auth.md:26]: `With` -> `We`(?)

File: adr/012-auth.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# 12. Authentication and Authorization
```
Comment (sfradkin) [inline adr/012-auth.md:1]: suggestion to also include a sequence diagram or flow diagram with this in order to better visualize the flow that is described.

File: adr/013-wrapped-domain-objects.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# 13. Wrapped Domain Objects
+
+Date: 2023-06-19
+
+## Decision
+
+We will "wrap" the HAPI FHIR (or any other FHIR library we may pivot to) resource classes with an interface to prevent the HAPI library from "poluting" our business logic.
```
Comment (sfradkin) [inline adr/013-wrapped-domain-objects.md:7]: `poluting` -> `polluting`

**Review by sfradkin — CHANGES_REQUESTED** [review]
ADR 13 needs a bit of editing as well.  I can post an change to the branch.

File: adr/012-auth.md (hunk 1 above)
Comment (halprin) [reply adr/012-auth.md:21]: Done.

File: adr/012-auth.md (hunk 2 above)
Comment (halprin) [reply adr/012-auth.md:24]: Done.

File: adr/012-auth.md (hunk 3 above)
Comment (halprin) [reply adr/012-auth.md:26]: Done.

File: adr/013-wrapped-domain-objects.md (hunk 5 above)
Comment (halprin) [reply adr/013-wrapped-domain-objects.md:7]: Done.

File: adr/012-auth.md (hunk 4 above)
Comment (halprin) [reply adr/012-auth.md:1]: That's a great idea.  I'll work on that.

**Review by jorg3lopez — APPROVED** [review]
Nice job!

**Review by sfradkin — APPROVED** [review]

**Review by sfradkin — APPROVED** [review]
I'm ok with adding the diagrams for ADR 12 as part of a different PR

**Review by JohnNKing — APPROVED** [review]
Looks good!

**Review by basiliskus — APPROVED** [review]

**Pull request merged**